pkg/declcfg/model_to_declcfg.go

```diff
@@ -0,0 +1,81 @@
+package declcfg
+
+import (
+	"sort"
+
+	"github.com/operator-framework/operator-registry/pkg/model"
+	"github.com/operator-framework/operator-registry/pkg/property"
+)
+
+func ConvertFromModel(mpkgs model.Model) DeclarativeConfig {
+	cfg := DeclarativeConfig{}
+	for _, mpkg := range mpkgs {
+		bundles := traverseModelChannels(*mpkg)
+
+		var i *Icon
+		if mpkg.Icon != nil {
+			i = &Icon{
+				Data:      mpkg.Icon.Data,
+				MediaType: mpkg.Icon.MediaType,
+			}
+		}
+		cfg.Packages = append(cfg.Packages, Package{
+			Schema:         schemaPackage,
+			Name:           mpkg.Name,
+			DefaultChannel: mpkg.DefaultChannel.Name,
+			Icon:           i,
+			Description:    mpkg.Description,
+		})
+		cfg.Bundles = append(cfg.Bundles, bundles...)
+	}
+
+	sort.Slice(cfg.Packages, func(i, j int) bool {
+		return cfg.Packages[i].Name < cfg.Packages[j].Name
+	})
+	sort.Slice(cfg.Bundles, func(i, j int) bool {
+		return cfg.Bundles[i].Name < cfg.Bundles[j].Name
+	})
+
+	return cfg
+}
+
+func traverseModelChannels(mpkg model.Package) []Bundle {
+	bundles := map[string]*Bundle{}
+
+	for _, ch := range mpkg.Channels {
+		for _, chb := range ch.Bundles {
+			b, ok := bundles[chb.Name]
+			if !ok {
+				b = &Bundle{
+					Schema:        schemaBundle,
+					Name:          chb.Name,
+					Package:       chb.Package.Name,
+					Image:         chb.Image,
+					RelatedImages: modelRelatedImagesToRelatedImages(chb.RelatedImages),
+					CsvJSON:       chb.CsvJSON,
+					Objects:       chb.Objects,
```

Comment:
If these are not set, for eg if CsvJSON and Objects for `chb` is not set during creating a bundle blob, won't this lead to a nil pointer dereferencing?

Reply:
This particular spot is okay as long as `chb` is non-nil. Which will always be the case unless there's a bug in our code that converts any of the representations to the model (or someone else is importing and using `model.Model` directly).

If `chb.CsvJSON` and/or `chb.Objects` is nil (definitely possible and happens frequently), then `declcfg` code should handle that. I'll check to see if the `Load` and `Write` function tests include cases for this (I think they do, but can't hurt to look again).

Reply:
@joelanford here's an annotations.yaml file for the bundle quay.io/operatorhubio/etcd:v0.6.1

```
annotations:
  operators.operatorframework.io.bundle.channel.default.v1: ""
  operators.operatorframework.io.bundle.channels.v1: alpha
  operators.operatorframework.io.bundle.manifests.v1: manifests/
  operators.operatorframework.io.bundle.mediatype.v1: registry+v1
  operators.operatorframework.io.bundle.metadata.v1: metadata/
  operators.operatorframework.io.bundle.package.v1: etcd
```

The default channel is not set here, which means when we're creating the model from that bundle, the `model.Package.DefaultChannel` is nil. 
Since we're accessing `mpkg.DefaultChannel.Name` here directly, it'll run into a nil pointer dereferencing error

Reply:
If the model, has been [validated](https://github.com/operator-framework/operator-registry/pull/601/files#diff-79ca10368c383e5ed5c7d8885fe1f4c2758b5ae4afd3feaf629e3a348b69b054R57-R59), `mpkg.DefaultChannel` won't be nil.

And (at least for operator-registry's uses of the model package) the conversion functions run the validations.
- [Declarative config to model](https://github.com/operator-framework/operator-registry/pull/601/files#diff-345bb90148fa16097184a5f10d48e80bf0cc6f8bfa5b6c7054b41070a50235aeR84-R86)
- [Sqlite to model](https://github.com/operator-framework/operator-registry/pull/601/files#diff-806b9455cc4618f4bd3e89d89ab1cb18499ec2e67f21b2914f8e2b69b057dc7eR28-R30)

Nonetheless, you're right that this code could result in a panic if used in other places or ways.  I think there are two possible solutions:
1. Add a `nil` check here for `mpkg.DefaultChannel` and if `nil`, use an empty string.
2. Run the full model validation at the beginning of the function, which would guarantee that `mpkg.DefaultChannel` would be non-nil.

With option 1, we're _kind of_ saying, that it's okay for this field to be unset, and as far as I know, we currently enforce that every package has a default channel (correct me if I'm wrong).

With option 2, we might be enforcing validations that are too strict.

As an aside, I think we should relax the model validations down to enforce only what the GRPC API requires.

Reply:
Fixed the `nil` pointer access issue with `model.Package.DefaultChannel.Name`.